### DISCUSSION TITLE
Welcome to zap Discussions!

### DISCUSSION THREAD

**Discussion by renerocksai** [zigzap/zap#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by edyu** [comment]
Hi Rene,
  I just finished a demo front-end written in Julia and my plan is to port it Zap as the foundation. I don't know enough about zig/zap to share ideas yet. On top of my head I'll need jinja2 templating so I'm looking at a c implementation of jinja2 to figure out how to port it to zig so I can use it with zap. Thank you for making it easy to use zap with the new zig packaging system.

See you at the conf.

**Comment by edyu** [comment]
There are many templating engines but I'll broadly categorize them into two categories:
  1. Less-verbose syntax to write HTML such as Mustache.
  2. Additional *escaped* control flow in HTML so that you repeat less such as jinja2.

As someone who prefer to see HTML/JavaScript directly, I prefer option 2. Hence the reason for jinja2 even though mustache is part of facil.io.

**Comment by renerocksai** [comment]
Hi Ed (I hope I got your name right),

Awesome! That's great to hear!

I love jinja2 templates, used them all the time in my python projects. But since I needed to make progress on zap, and facil.io already had mustache templating, I decided that it was easier to just zig-wrap the mustache stuff than trying to integrate another codebase. I checked that mustache supported all the primitives that I needed for conditionals and loops in my JSON templates, which further confirmed me in that decision.

But if you manage to ziggify a jinja2 templating codebase, that would be awesome! I would integrate it into zap but I guess it would be more valuable to other projects as a standalone lib.

Another reason why I chose the built-in mustache code: I don't need to make copies of string slices. Many C libs suffer from the problem that they use C strings which require 0-termination. When my string is a view into a buffer, there is no 0-termination byte at the end. So I'd need to make a copy that is 1 byte longer than my slice and contains the trailing 0.

Watch out for that. It's not the end of the world but I presume that template-heavy apps might take a significant performance hit from copying strings around all the time, depending on what else they do. However, if the alternative would be: no jinja2, I would happily take the perf hit and figure out how to modify the C code later.

Looking forward to seeing you at the conference, too!

-Rene

**Reply by edyu** [reply]
I might actually be wrong. Mustache might be even better for my current use case. Maybe I mixed it with other html shorthand template engines.
